### PR TITLE
update database schemas for sites, forecasts, observations

### PR DESCRIPTION
**Pull request by alorenzo175**
after #52 is merged

### CODE REVIEW

**Comment by alorenzo175** [comment]
I've set many metadata fields to DECIMAL to avoid float conversion changing values. Are the precision of the DECIMAL fields enough? DECIMAL(10, 6) means there will be 6 digits after the decimal and 4 before. Did I miss any parameters?  @wholmgren

File: datastore/migrations/0003_add_other_tables.up.sql [hunk 1]
```diff
@@ -3,7 +3,27 @@ CREATE TABLE arbiter_data.sites (
     id BINARY(16) NOT NULL DEFAULT (UUID_TO_BIN(UUID(), 1)),
     organization_id BINARY(16) NOT NULL,
     name VARCHAR(64) NOT NULL,
-    -- add everything else later...
+    latitude DECIMAL(8, 6) NOT NULL,
+    longitude DECIMAL(9, 6) NOT NULL,
```
Comment (wholmgren) [inline datastore/migrations/0003_add_other_tables.up.sql]: negative ok in this spec?

File: datastore/migrations/0003_add_other_tables.up.sql [hunk 2]
```diff
@@ -3,7 +3,27 @@ CREATE TABLE arbiter_data.sites (
...
+    elevation DECIMAL(7, 2) NOT NULL,
+    timezone VARCHAR(32) NOT NULL,
+    extra_parameters VARCHAR(255) NOT NULL,
+    ac_capacity DECIMAL(10, 5) NOT NULL,
+    dc_capacity DECIMAL(10, 5) NOT NULL,
```
Comment (wholmgren) [inline datastore/migrations/0003_add_other_tables.up.sql]: 0.00001 MW to 99999.99999 MW seems ok. add a digit and no one would ever complain.

File: datastore/migrations/0003_add_other_tables.up.sql [hunk 3]
```diff
@@ -18,8 +38,14 @@ CREATE TABLE arbiter_data.observations (
...
     site_id BINARY(16) NOT NULL,
     name VARCHAR(64) NOT NULL,
-
-    -- add everything else later...
+    variable VARCHAR(32) NOT NULL,
+    interval_label ENUM('beginning', 'ending', 'instant') NOT NULL,
+    interval_length SMALLINT UNSIGNED NOT NULL,
+    value_type VARCHAR(32) NOT NULL,
+    uncertainty FLOAT NOT NULL,
+    extra_parameters VARCHAR(255) NOT NULL,
```
Comment (wholmgren) [inline datastore/migrations/0003_add_other_tables.up.sql]: only 255 characters? I can imagine a verbose metadata spec bumping into that

File: datastore/migrations/0003_add_other_tables.up.sql (hunk 1 above)
Comment (alorenzo175) [reply datastore/migrations/0003_add_other_tables.up.sql]: yes, all decimal fields are +- the range

File: datastore/migrations/0003_add_other_tables.up.sql (hunk 3 above)
Comment (wholmgren) [reply datastore/migrations/0003_add_other_tables.up.sql]: more likely for a site than an obs, commented in the wrong place

File: datastore/migrations/0003_add_other_tables.up.sql [hunk 4]
```diff
@@ -18,8 +38,14 @@ CREATE TABLE arbiter_data.observations (
     organization_id BINARY(16) NOT NULL,
     site_id BINARY(16) NOT NULL,
     name VARCHAR(64) NOT NULL,
-
-    -- add everything else later...
+    variable VARCHAR(32) NOT NULL,
```
Comment (wholmgren) [inline datastore/migrations/0003_add_other_tables.up.sql]: did you consider `enum` here? I think enum would prove to be overly restrictive, but it's consistent with the current data model

File: datastore/migrations/0003_add_other_tables.up.sql [hunk 5]
```diff
@@ -3,7 +3,27 @@ CREATE TABLE arbiter_data.sites (
...
+    temperature_coefficient DECIMAL(7, 5) NOT NULL,
+    tracking_type ENUM('fixed', 'single_axis') NOT NULL,
+    surface_tilt DECIMAL(4, 2),
+    surface_azimuth DECIMAL(5, 2),
+    axis_tilt DECIMAL(4, 2),
+    axis_azimuth DECIMAL(5, 2),
+    ground_coverage_ratio DECIMAL(8, 4),
+    backtrack BOOLEAN,
+    max_rotation_angle DECIMAL(5, 2),
+    irradiance_loss_factor DECIMAL(5, 2) NOT NULL,
```
Comment (wholmgren) [inline datastore/migrations/0003_add_other_tables.up.sql]: not supporting this

File: datastore/tests/conftest.py [hunk 6]
```diff
@@ -95,16 +96,33 @@ def fcn(action, object_type, applies_to_all, org=None):
...
-            'INSERT INTO sites (id, organization_id, name) VALUES '
-            '(%s, %s, %s)', list(out.values()))
+                          name=f'site{str(uuid1())[:10]}',
+                          latitude=0, longitude=0, elevation=0,
+                          timezone='America/Denver', extra_parameters='',
+                          ac_capacity=0, dc_capacity=0,
+                          temperature_coefficient=0,
+                          tracking_type='fixed',
+                          surface_tilt=0, surface_azimuth=0,
+                          axis_tilt=0, axis_azimuth=0,
```
Comment (wholmgren) [inline datastore/tests/conftest.py]: if `tracking_type='fixed'` should the single axis parameters be `NULL`? Is that allowed in the schema?

File: datastore/migrations/0003_add_other_tables.up.sql (hunk 4 above)
Comment (alorenzo175) [reply datastore/migrations/0003_add_other_tables.up.sql]: considered but didn't want to have to update the schema if we add things later

File: datastore/tests/conftest.py (hunk 6 above)
Comment (alorenzo175) [reply datastore/tests/conftest.py]: yes, and it is allowed

Comment (alorenzo175) [reply datastore/tests/conftest.py]: overall, want to try and avoid allowing a column to be null because it requires extra space

Comment (wholmgren) [reply datastore/tests/conftest.py]: space as in memory? Aren't we talking about a few bytes per site? Seems worthwhile to promote internal data model consistency.

File: datastore/migrations/0003_add_other_tables.up.sql [hunk 7]
```diff
@@ -3,7 +3,26 @@ CREATE TABLE arbiter_data.sites (
     id BINARY(16) NOT NULL DEFAULT (UUID_TO_BIN(UUID(), 1)),
     organization_id BINARY(16) NOT NULL,
     name VARCHAR(64) NOT NULL,
-    -- add everything else later...
+    latitude DECIMAL(8, 6) NOT NULL,
+    longitude DECIMAL(9, 6) NOT NULL,
+    elevation DECIMAL(7, 2) NOT NULL,
+    timezone VARCHAR(32) NOT NULL,
+    extra_parameters TEXT NOT NULL,  -- unclear if this will have a positive or negative effect on performance
```
Comment (wholmgren) [inline datastore/migrations/0003_add_other_tables.up.sql]: I thought a limit was reasonable, just something bigger than 255 characters. We should put a limit in the API.

File: datastore/tests/conftest.py (hunk 6 above)
Comment (alorenzo175) [reply datastore/tests/conftest.py]: I meant avoid null in general, but these parameters are allowed to be null to be consistent. Space as in memory and disk.

Comment (alorenzo175) [reply datastore/tests/conftest.py]: and define then (now as None) in conftest for easier comparison to result from DB later

File: datastore/migrations/0003_add_other_tables.up.sql (hunk 7 above)
Comment (alorenzo175) [reply datastore/migrations/0003_add_other_tables.up.sql]: I agree about an API limit, but what limit is reasonable? 
```
This webpage documents the public Solar Forecast Arbiter API. This RESTful API is primarily meant to be accessed by the Solar Forecast Arbiter dashboard. The API relies primarily on JSON data structures in requests and responses, with the notable exception of the forecast and observation GET/POST data endpoints which also support CSV files
```
is already 341 characters

Comment (alorenzo175) [reply datastore/migrations/0003_add_other_tables.up.sql]: I think it's ok to let the api set the limit and store whatever it gives to the db

**Pull request merged**